File: leetcode/1866.go
package leetcode

func rearrangeSticks(n int, k int) int {
	dp := make([]int, k+1)
	dp[0] = 1
	for i := 0; i < n; i++ {
		for j := k; j > 0; j-- {
			dp[j] = (i*dp[j] + dp[j-1]) % 1000000007
		}
		dp[0] = 0
	}
	return dp[k]
}

File: leetcode/1866.cpp
class Solution {
    const int mod = 1e9 + 7;

public:
    int rearrangeSticks(int n, int k) {
        vector<int> f(k + 1);
        f[0] = 1;
        for (int i = 1; i <= n; ++i) {
            for (int j = min(k, i); j > 0; --j) {
                f[j] = ((long long)f[j] * (i - 1) % mod + f[j - 1]) % mod;
            }
            f[0] = 0;
        }
        return f[k];
    }
};